File: backend/scripts/corpus-fetch.ts
import "dotenv/config";
import fs from "node:fs/promises";
import path from "node:path";

type Source = {
  kind: "ecfr-xml" | "pdf";
  url: string;
  out_filename: string;
  doc_class: string;
  doc_id: string;
  doc_title: string;
};

const here = path.resolve(__dirname, "..");
const manifestPath = path.join(here, "corpus-sources", "sources.json");
const rawDir = path.join(here, "corpus-sources", "raw");

const force = process.argv.includes("--force");

(async () => {
  const sources: Source[] = JSON.parse(await fs.readFile(manifestPath, "utf8"));
  await fs.mkdir(rawDir, { recursive: true });

  for (const s of sources) {
    const out = path.join(rawDir, s.out_filename);
    let exists = false;
    try {
      const st = await fs.stat(out);
      exists = st.size > 0;
    } catch {}
    if (exists && !force) {
      console.log(`skip  ${s.out_filename} (exists; --force to refetch)`);
      continue;
    }
    process.stdout.write(`fetch ${s.out_filename} … `);
    const r = await fetch(s.url, {
      headers: { Accept: s.kind === "ecfr-xml" ? "application/xml" : "application/pdf" },
    });
    if (!r.ok) {
      console.log(`FAILED (${r.status} ${r.statusText})`);
      console.error(`  url: ${s.url}`);
      process.exitCode = 1;
      continue;
    }
    const buf = Buffer.from(await r.arrayBuffer());
    await fs.writeFile(out, buf);
    console.log(`ok (${(buf.length / 1024).toFixed(0)} KB)`);
  }

  console.log("corpus-fetch: done.");
})().catch((e) => {
  console.error(e);
  process.exit(1);
});

File: backend/scripts/corpus_fetch.py
"""Download corpus source documents per backend/corpus-sources/sources.json."""

from __future__ import annotations

import asyncio
import json
import sys
from pathlib import Path

import httpx

HERE = Path(__file__).resolve().parent.parent
MANIFEST = HERE / "corpus-sources" / "sources.json"
RAW_DIR = HERE / "corpus-sources" / "raw"


async def main() -> None:
    sources = json.loads(MANIFEST.read_text("utf-8"))
    RAW_DIR.mkdir(parents=True, exist_ok=True)
    force = "--force" in sys.argv

    async with httpx.AsyncClient(timeout=60.0, follow_redirects=True) as client:
        for s in sources:
            out = RAW_DIR / s["out_filename"]
            if out.exists() and out.stat().st_size > 0 and not force:
                print(f"skip  {s['out_filename']} (exists; --force to refetch)")
                continue
            accept = "application/xml" if s["kind"] == "ecfr-xml" else "application/pdf"
            print(f"fetch {s['out_filename']} … ", end="", flush=True)
            r = await client.get(s["url"], headers={"Accept": accept})
            if r.status_code != 200:
                print(f"FAILED ({r.status_code})")
                print(f"  url: {s['url']}")
                continue
            out.write_bytes(r.content)
            print(f"ok ({len(r.content) / 1024:.0f} KB)")

    print("corpus-fetch: done.")


if __name__ == "__main__":
    asyncio.run(main())
